File: tests/oracle/external_stub_1204.cc
/*
 * EXTERNAL-STUB-1204: locked Ghidra 12.0.4 construction-period space
 * registration oracle (EXTERNAL-stub support wave).
 *
 * The fixture drives the real AddrSpaceManager decode path that production
 * architectures use to register spaces from a spec —
 * AddrSpaceManager::decodeSpaces (translate.cc:281-303, including the
 * DEFAULT-space registration via the `defaultspace` attribute) and
 * AddrSpaceManager::decodeSpace (translate.cc:254-275, the element-id
 * dispatch to the partial SpacebaseSpace/UniqueSpace/OtherSpace/
 * OverlaySpace/AddrSpace constructors) — over a hand-built Element tree,
 * plus the readSpace resolution errors (marshal.cc:400-409) and the
 * EXTERNAL-named processor-space registration the Ghidra platform side
 * defines for import stubs (Java AddressSpace.java:80; the 12.0.4
 * decompiler oracle has no ExternalSpace — see the fixture's Rust twin).
 * Every observation is printed in the shared line format so the Rust
 * comparand must match byte for byte.
 */

#include <bits/stdc++.h>

// Test-only access is required to reach the protected decode entry points
// (production reaches them from Translate subclasses during initialization)
// and to read AddrSpace::refcount. This matches the access-hack precedent
// of the space_registry fixture and is confined to this translation unit.
#define class struct
#define private public
#include "address.hh"
#include "fspec.hh"
#include "op.hh"
#include "space.hh"
#include "translate.hh"
#undef private
#undef class

using namespace ghidra;

class FixtureTranslate final : public Translate {
public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
  }

  // AddrSpaceManager protected bridge, exactly like production Translate
  // subclasses call these during initialization.
  std::string tryDecodeSpaces(XmlDecode &decoder) {
    try {
      decodeSpaces(decoder, this);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
    catch (DecoderError &err) {
      return std::string("err ") + err.explain;
    }
  }

  std::string tryDecodeSpace(AddrSpace *&out, XmlDecode &decoder) {
    try {
      out = decodeSpace(decoder, this);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
    catch (DecoderError &err) {
      return std::string("err ") + err.explain;
    }
  }

  std::string tryInsertSpace(AddrSpace *spc) {
    try {
      insertSpace(spc);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    static VarnodeData dummy;
    return dummy;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

// Helpers building the hand-parsed Element tree: a <spaces> document with
// <space>/<space_unique>/<space_base>/<space_overlay> children, exactly the
// shapes SleighArchitecture feeds decodeSpaces in production.
static Element *makeSpaceElement(Element *parent, const std::string &tag,
                                 const std::vector<std::pair<std::string,
                                 std::string>> &attributes) {
  Element *el = new Element(parent);
  el->setName(tag);
  for (size_t i = 0; i < attributes.size(); ++i)
    el->addAttribute(attributes[i].first, attributes[i].second);
  parent->addChild(el);
  return el;
}

static Element *makeSpacesDocument(const std::string &defaultspace) {
  Element *spaces = new Element((Element *)0);
  spaces->setName("spaces");
  spaces->addAttribute("defaultspace", defaultspace);
  return spaces;
}

static std::string hexU64(uintb v) {
  std::ostringstream s;
  s << "0x" << std::hex << v;
  return s.str();
}

static void printSpaceLine(const AddrSpace *spc) {
  std::cout << "  space idx=" << spc->getIndex()
            << " name=" << spc->getName()
            << " type=" << (int4)spc->getType()
            << " addrsize=" << spc->getAddrSize()
            << " wordsize=" << spc->getWordSize()
            << " endian=" << (spc->isBigEndian() ? "big" : "l")
            << " shortcut=" << spc->getShortcut()
            << " delay=" << spc->getDelay()
            << " dead=" << spc->getDeadcodeDelay()
            << " highest=" << hexU64(spc->getHighest())
            << " plb=" << hexU64(spc->getPointerLowerBound())
            << " pub=" << hexU64(spc->getPointerUpperBound())
            << " ref=" << spc->refcount
            << " h=" << (spc->isHeritaged() ? 1 : 0)
            << " dc=" << (spc->doesDeadcode() ? 1 : 0)
            << " fs=" << (spc->isFormalStackSpace() ? 1 : 0)
            << " ov=" << (spc->isOverlay() ? 1 : 0)
            << " ob=" << (spc->isOverlayBase() ? 1 : 0)
            << " hp=" << (spc->hasPhysical() ? 1 : 0)
            << " oo=" << (spc->isOtherSpace() ? 1 : 0)
            << std::endl;
}

static void printWalk(const AddrSpaceManager *mgr) {
  std::cout << "  walk";
  for (AddrSpace *spc = mgr->getNextSpaceInOrder((AddrSpace *)0);
       spc != (AddrSpace *)0 && spc != (AddrSpace *)~((uintp)0);
       spc = mgr->getNextSpaceInOrder(spc)) {
    std::cout << " " << spc->getName();
  }
  std::cout << std::endl;
}

// Canonical decodeSpaces input: const is implicit (decodeSpaces inserts it
// itself), unique=2, ram=3, register=4, default=ram.
static Element *canonicalSpaces(void) {
  Element *spaces = makeSpacesDocument("ram");
  makeSpaceElement(spaces, "space_unique",
                   {{"name", "unique"}, {"index", "2"}, {"size", "4"},
                    {"delay", "0"}});
  makeSpaceElement(spaces, "space",
                   {{"name", "ram"}, {"index", "3"}, {"size", "8"},
                    {"delay", "0"}, {"physical", "true"}});
  makeSpaceElement(spaces, "space",
                   {{"name", "register"}, {"index", "4"}, {"size", "8"},
                    {"delay", "0"}, {"physical", "true"}});
  return spaces;
}

int main(void) {
  // The id lookups (AttributeId::find/ElementId::find) are filled by the
  // explicit initialize calls that every production entry point makes
  // (libdecomp.cc:23-24); the fixture must do the same before decoding.
  AttributeId::initialize();
  ElementId::initialize();
  std::cout << "schema=1|fixture=EXTERNAL-STUB-1204|"
               "oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
            << std::endl;

  // ---- case 1: decodeSpaces canonical registration + DEFAULT ----------
  std::cout << "case=decode_spaces_default_registration" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = canonicalSpaces();
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    std::cout << "  numSpaces=" << tr.numSpaces() << std::endl;
    for (int4 i = 0; i < tr.numSpaces(); ++i)
      if (tr.getSpace(i) != (AddrSpace *)0)
        printSpaceLine(tr.getSpace(i));
    std::cout << "  defaultSize=" << tr.getDefaultSize()
              << " code=" << tr.getDefaultCodeSpace()->getName()
              << " data=" << tr.getDefaultDataSpace()->getName() << std::endl;
    printWalk(&tr);
    delete doc;
  }

  // ---- case 2: <space_overlay> decode marks the base -------------------
  std::cout << "case=decode_space_overlay_marks_base" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = canonicalSpaces();
    makeSpaceElement(doc, "space_overlay",
                     {{"name", "ov"}, {"index", "5"}, {"base", "ram"}});
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    AddrSpace *ov = tr.getSpaceByName("ov");
    AddrSpace *ram = tr.getSpaceByName("ram");
    std::cout << "  ram_ob=" << (ram->isOverlayBase() ? 1 : 0)
              << " ov_ov=" << (ov->isOverlay() ? 1 : 0)
              << " contain=" << ov->getContain()->getName() << std::endl;
    printSpaceLine(ov);
    delete doc;
  }

  // ---- case 3: <space_base> decode resolves contain --------------------
  std::cout << "case=decode_space_base_stack_contain" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = canonicalSpaces();
    makeSpaceElement(doc, "space_base",
                     {{"name", "stack"}, {"index", "5"}, {"size", "8"},
                      {"delay", "1"}, {"contain", "ram"}});
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    AddrSpace *stack = tr.getSpaceByName("stack");
    std::cout << "  stackSlot=" << tr.getStackSpace()->getName()
              << " type=" << (int4)stack->getType()
              << " contain=" << stack->getContain()->getName()
              << " numBase=" << stack->numSpacebase()
              << " growsNeg=" << (stack->stackGrowsNegative() ? 1 : 0)
              << " formal=" << (stack->isFormalStackSpace() ? 1 : 0)
              << std::endl;
    printSpaceLine(stack);
    delete doc;
  }

  // ---- case 4: readSpace resolution failure ----------------------------
  std::cout << "case=decode_space_unknown_base" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = canonicalSpaces();
    makeSpaceElement(doc, "space_overlay",
                     {{"name", "bad"}, {"index", "6"}, {"base", "nope"}});
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    std::cout << "  ovLookup="
              << (tr.getSpaceByName("bad") == (AddrSpace *)0
                      ? std::string("null")
                      : tr.getSpaceByName("bad")->getName())
              << std::endl;
    delete doc;
  }

  // ---- case 5: bad defaultspace attribute ------------------------------
  std::cout << "case=decode_spaces_bad_default" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = makeSpacesDocument("missing");
    makeSpaceElement(doc, "space",
                     {{"name", "ram"}, {"index", "3"}, {"size", "8"},
                      {"delay", "0"}});
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    std::cout << "  default="
              << (tr.getDefaultCodeSpace() == (AddrSpace *)0
                      ? std::string("null")
                      : tr.getDefaultCodeSpace()->getName())
              << std::endl;
    delete doc;
  }

  // ---- case 6: EXTERNAL-named processor space registration -------------
  // The Ghidra platform side defines the external space as a flat 32-bit
  // space named "EXTERNAL" (AddressSpace.java:80) and the ELF importer
  // materializes import stubs in a default-space EXTERNAL memory block.
  // The 12.0.4 decompiler registers it like any other processor space:
  // no overlay flags, no special type handling.
  std::cout << "case=external_named_space_registration" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = canonicalSpaces();
    XmlDecode setup(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(setup) << std::endl;
    AddrSpace *external = new AddrSpace(&tr, &tr, IPTR_PROCESSOR, "EXTERNAL",
                                        false, 4, 1, 9, 0, 0, 0);
    std::cout << "  " << tr.tryInsertSpace(external) << std::endl;
    AddrSpace *ram = tr.getSpaceByName("ram");
    std::cout << "  lookup="
              << (tr.getSpaceByName("EXTERNAL") == (AddrSpace *)0
                      ? std::string("null")
                      : tr.getSpaceByName("EXTERNAL")->getName())
              << " shortcut=" << external->getShortcut()
              << " ram_ob=" << (ram->isOverlayBase() ? 1 : 0) << std::endl;
    printSpaceLine(external);
    AddrSpace *duplicate = new AddrSpace(&tr, &tr, IPTR_PROCESSOR, "EXTERNAL",
                                         false, 4, 1, 10, 0, 0, 0);
    std::cout << "  " << tr.tryInsertSpace(duplicate) << std::endl;
    printWalk(&tr);
    delete doc;
  }

  // ---- case 7: deadcodedelay default from delay ------------------------
  std::cout << "case=deadcodedelay_default_from_delay" << std::endl;
  {
    FixtureTranslate tr;
    Element *doc = makeSpacesDocument("ws");
    makeSpaceElement(doc, "space",
                     {{"name", "ws"}, {"index", "3"}, {"size", "4"},
                      {"wordsize", "2"}, {"delay", "3"}});
    XmlDecode decoder(&tr, doc);
    std::cout << "  " << tr.tryDecodeSpaces(decoder) << std::endl;
    AddrSpace *ws = tr.getSpaceByName("ws");
    printSpaceLine(ws);
    FixtureTranslate tr2;
    Element *doc2 = makeSpacesDocument("ws2");
    makeSpaceElement(doc2, "space",
                     {{"name", "ws2"}, {"index", "3"}, {"size", "4"},
                      {"delay", "2"}, {"deadcodedelay", "5"}});
    XmlDecode decoder2(&tr2, doc2);
    std::cout << "  " << tr2.tryDecodeSpaces(decoder2) << std::endl;
    AddrSpace *ws2 = tr2.getSpaceByName("ws2");
    printSpaceLine(ws2);
    delete doc;
    delete doc2;
  }
  return 0;
}

File: tests/oracle/external_stub_1204.rs
//! EXTERNAL-STUB-1204 Rust comparand: construction-period space
//! registration through `SpaceRegistry::decode_spaces`/`decode_space`
//! (the ports of translate.cc:281/254), the EXTERNAL-named space
//! registration, and the decode-path error surfaces. Output must match the
//! locked Ghidra 12.0.4 fixture byte for byte
//! (`tests/oracle/external_stub_1204.cc`).

use std::sync::{Arc, RwLock};

use rugra::marshal::{Element, IdRegistry, TreeDecoder};
use rugra::space::{AddrSpace, SpaceType, SpaceRegistry};

fn hex_u64(v: u64) -> String {
    format!("{:#x}", v)
}

fn print_space_line(spc: &AddrSpace) {
    println!(
        "  space idx={} name={} type={} addrsize={} wordsize={} endian={} shortcut={} delay={} dead={} highest={} plb={} pub={} ref={} h={} dc={} fs={} ov={} ob={} hp={} oo={}",
        spc.get_index(),
        spc.get_name(),
        spc.get_type() as i32,
        spc.get_addr_size(),
        spc.get_word_size(),
        if spc.is_big_endian() { "big" } else { "l" },
        spc.get_shortcut(),
        spc.get_delay(),
        spc.get_deadcode_delay(),
        hex_u64(spc.get_highest()),
        hex_u64(spc.get_pointer_lower_bound()),
        hex_u64(spc.get_pointer_upper_bound()),
        spc.refcount(),
        i32::from(spc.is_heritaged()),
        i32::from(spc.does_deadcode()),
        i32::from(spc.is_formal_stackspace()),
        i32::from(spc.is_overlay()),
        i32::from(spc.is_overlay_base()),
        i32::from(spc.has_physical()),
        i32::from(spc.is_other_space()),
    );
}

fn print_walk(registry: &SpaceRegistry) {
    print!("  walk");
    let mut cursor = registry.get_next_space_in_order(None);
    while let Some(spc) = cursor {
        print!(" {}", spc.get_name());
        cursor = registry.get_next_space_in_order(Some(spc));
    }
    println!();
}

fn make_space_element(
    parent: &mut Element,
    tag: &str,
    attributes: &[(&str, &str)],
) -> Arc<RwLock<Element>> {
    let mut child = Element::new();
    child.set_name(tag);
    for (name, value) in attributes {
        child.add_attribute(name, value);
    }
    let child = Arc::new(RwLock::new(child));
    parent.add_child(Arc::clone(&child));
    child
}

fn make_spaces_document(defaultspace: &str) -> Element {
    let mut spaces = Element::new();
    spaces.set_name("spaces");
    spaces.add_attribute("defaultspace", defaultspace);
    spaces
}

fn canonical_spaces() -> Element {
    let mut spaces = make_spaces_document("ram");
    make_space_element(
        &mut spaces,
        "space_unique",
        &[("name", "unique"), ("index", "2"), ("size", "4"), ("delay", "0")],
    );
    make_space_element(
        &mut spaces,
        "space",
        &[
            ("name", "ram"),
            ("index", "3"),
            ("size", "8"),
            ("delay", "0"),
            ("physical", "true"),
        ],
    );
    make_space_element(
        &mut spaces,
        "space",
        &[
            ("name", "register"),
            ("index", "4"),
            ("size", "8"),
            ("delay", "0"),
            ("physical", "true"),
        ],
    );
    spaces
}

fn decode_spaces(registry: &mut SpaceRegistry, doc: Element) -> String {
    let root = Arc::new(RwLock::new(doc));
    let id_registry = Arc::new(RwLock::new(IdRegistry::new()));
    let mut decoder = TreeDecoder::new(root, id_registry);
    match registry.decode_spaces(&mut decoder) {
        Ok(()) => "ok".to_string(),
        Err(message) => format!("err {}", message),
    }
}

fn main() {
    println!(
        "schema=1|fixture=EXTERNAL-STUB-1204|oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
    );

    // ---- case 1: decodeSpaces canonical registration + DEFAULT ----------
    println!("case=decode_spaces_default_registration");
    {
        let mut registry = SpaceRegistry::new();
        let doc = canonical_spaces();
        println!("  {}", decode_spaces(&mut registry, doc));
        println!("  numSpaces={}", registry.num_spaces());
        for index in 0..registry.num_spaces() {
            if let Some(spc) = registry.get_space(index) {
                print_space_line(&spc);
            }
        }
        println!(
            "  defaultSize={} code={} data={}",
            registry.get_default_size(),
            registry
                .get_default_code_space()
                .map(|s| s.get_name())
                .unwrap_or_default(),
            registry
                .get_default_data_space()
                .map(|s| s.get_name())
                .unwrap_or_default(),
        );
        print_walk(&registry);
    }

    // ---- case 2: <space_overlay> decode marks the base -------------------
    println!("case=decode_space_overlay_marks_base");
    {
        let mut registry = SpaceRegistry::new();
        let mut doc = canonical_spaces();
        make_space_element(
            &mut doc,
            "space_overlay",
            &[("name", "ov"), ("index", "5"), ("base", "ram")],
        );
        println!("  {}", decode_spaces(&mut registry, doc));
        let ov = registry.get_space_by_name("ov").expect("ov registered");
        let ram = registry.get_space_by_name("ram").expect("ram registered");
        println!(
            "  ram_ob={} ov_ov={} contain={}",
            i32::from(ram.is_overlay_base()),
            i32::from(ov.is_overlay()),
            ov.get_contain().map(|s| s.get_name()).unwrap_or_default(),
        );
        print_space_line(&ov);
    }

    // ---- case 3: <space_base> decode resolves contain --------------------
    println!("case=decode_space_base_stack_contain");
    {
        let mut registry = SpaceRegistry::new();
        let mut doc = canonical_spaces();
        make_space_element(
            &mut doc,
            "space_base",
            &[
                ("name", "stack"),
                ("index", "5"),
                ("size", "8"),
                ("delay", "1"),
                ("contain", "ram"),
            ],
        );
        println!("  {}", decode_spaces(&mut registry, doc));
        let stack = registry.get_space_by_name("stack").expect("stack registered");
        println!(
            "  stackSlot={} type={} contain={} numBase={} growsNeg={} formal={}",
            registry
                .get_stack_space()
                .map(|s| s.get_name())
                .unwrap_or_default(),
            stack.get_type() as i32,
            stack
                .get_contain()
                .map(|s| s.get_name())
                .unwrap_or_default(),
            stack.num_spacebase(),
            i32::from(stack.stack_grows_negative()),
            i32::from(stack.is_formal_stackspace()),
        );
        print_space_line(&stack);
    }

    // ---- case 4: readSpace resolution failure ----------------------------
    println!("case=decode_space_unknown_base");
    {
        let mut registry = SpaceRegistry::new();
        let mut doc = canonical_spaces();
        make_space_element(
            &mut doc,
            "space_overlay",
            &[("name", "bad"), ("index", "6"), ("base", "nope")],
        );
        println!("  {}", decode_spaces(&mut registry, doc));
        println!(
            "  ovLookup={}",
            registry
                .get_space_by_name("bad")
                .map(|s| s.get_name())
                .unwrap_or_else(|| "null".to_string()),
        );
    }

    // ---- case 5: bad defaultspace attribute ------------------------------
    println!("case=decode_spaces_bad_default");
    {
        let mut registry = SpaceRegistry::new();
        let mut doc = make_spaces_document("missing");
        make_space_element(
            &mut doc,
            "space",
            &[("name", "ram"), ("index", "3"), ("size", "8"), ("delay", "0")],
        );
        println!("  {}", decode_spaces(&mut registry, doc));
        println!(
            "  default={}",
            registry
                .get_default_code_space()
                .map(|s| s.get_name())
                .unwrap_or_else(|| "null".to_string()),
        );
    }

    // ---- case 6: EXTERNAL-named processor space registration -------------
    println!("case=external_named_space_registration");
    {
        let mut registry = SpaceRegistry::new();
        let doc = canonical_spaces();
        println!("  {}", decode_spaces(&mut registry, doc));
        // The Ghidra-platform external space (AddressSpace.java:80): flat
        // 32-bit space named EXTERNAL, registered like any processor space.
        let external = AddrSpace::new_external_space(9, false);
        match registry.insert_space(external.clone()) {
            Ok(()) => println!("  ok"),
            Err(message) => println!("  err {}", message),
        }
        let ram = registry.get_space_by_name("ram").expect("ram registered");
        println!(
            "  lookup={} shortcut={} ram_ob={}",
            registry
                .get_space_by_name("EXTERNAL")
                .map(|s| s.get_name())
                .unwrap_or_else(|| "null".to_string()),
            external.get_shortcut(),
            i32::from(ram.is_overlay_base()),
        );
        print_space_line(&external);
        let duplicate = AddrSpace::new_external_space(10, false);
        match registry.insert_space(duplicate) {
            Ok(()) => println!("  ok"),
            Err(message) => println!("  err {}", message),
        }
        print_walk(&registry);
    }

    // ---- case 7: deadcodedelay default from delay ------------------------
    println!("case=deadcodedelay_default_from_delay");
    {
        let mut registry = SpaceRegistry::new();
        let mut doc = make_spaces_document("ws");
        make_space_element(
            &mut doc,
            "space",
            &[
                ("name", "ws"),
                ("index", "3"),
                ("size", "4"),
                ("wordsize", "2"),
                ("delay", "3"),
            ],
        );
        println!("  {}", decode_spaces(&mut registry, doc));
        let ws = registry.get_space_by_name("ws").expect("ws registered");
        print_space_line(&ws);

        let mut registry2 = SpaceRegistry::new();
        let mut doc2 = make_spaces_document("ws2");
        make_space_element(
            &mut doc2,
            "space",
            &[
                ("name", "ws2"),
                ("index", "3"),
                ("size", "4"),
                ("delay", "2"),
                ("deadcodedelay", "5"),
            ],
        );
        println!("  {}", decode_spaces(&mut registry2, doc2));
        let ws2 = registry2.get_space_by_name("ws2").expect("ws2 registered");
        print_space_line(&ws2);
    }
    let _ = SpaceType::Processor;
}
